bulk-user-export-tool/scim-bulk-user-export-tool/src/main/java/org/wso2/scim/bulk/export/BulkExportUsers.java

```diff
@@ -80,68 +87,107 @@ public static void main(String args[]) throws IOException, KeyStoreException,
+                    stringResponse = EntityUtils.toString(response.getEntity(), "UTF-8");
+                    JsonNode jsonTree = new ObjectMapper().readTree(stringResponse);
+                    JsonNode usersNode = jsonTree.at("/Resources");
+
+                    // Create a flattened user information array.
+                    if (usersNode.isArray()) {
+                        ArrayNode arrayNode = (ArrayNode) usersNode;
+                        for (int i = 0; i < arrayNode.size(); i++) {
+                            JsonNode arrayElement = arrayNode.get(i);
+                            usersArrayNode.add(JSONFlattener.generateFlatJSON(new ObjectMapper().createObjectNode(),
+                                    arrayElement, null, Collections.emptySet()));
+                        }
+                        if (arrayNode.size() < count) {
+                            LOGGER.log(Level.INFO, "End of results reached.");
+                            break;
+                        }
+                    } else {
+                        LOGGER.log(Level.INFO, "End of results reached.");
+                        break;
                     }
-                }
 
-                // Create a csv schema and generate the csv file containing user information.
-                CsvSchema.Builder csvSchemaBuilder = CsvSchema.builder();
-                for (int i = 0; i < usersArrayNode.size(); i++) {
-                    usersArrayNode.get(i).fieldNames().forEachRemaining(
-                            fieldName -> {
-                                if (!csvSchemaBuilder.hasColumn(fieldName)) {
-                                    csvSchemaBuilder.addColumn(fieldName);
-                                }
-                            });
+                } else {
+                    LOGGER.log(Level.INFO, request.getMethod() + " Request to " + request.getURI().toString() +
+                            " returned the status code : " + response.getStatusLine());
+                    return;
                 }
-                CsvSchema csvSchema = csvSchemaBuilder.build().withHeader();
-
-                CsvMapper csvMapper = new CsvMapper();
-                csvMapper.writerFor(ArrayNode.class)
-                        .with(csvSchema)
-                        .writeValue(new File(csvDirectory), usersArrayNode);
-                LOGGER.log(Level.INFO, "User information was successfully written to : " + csvDirectory + " file.");
-            } else {
-                LOGGER.log(Level.INFO, request.getMethod() + " Request to " + request.getURI().toString() +
-                        " returned the status code : " + response.getStatusLine());
             }
         }
+
+        // Create a csv schema and generate the csv file containing user information.
+        CsvSchema.Builder csvSchemaBuilder = CsvSchema.builder();
+        for (int i = 0; i < usersArrayNode.size(); i++) {
```

Comment:
Related to https://github.com/wso2/samples-is/pull/285#issuecomment-850611008

The entire use case is loaded into a single ArrayNode called usersArrayNode and the CSV is created after that. This could lead to memory issues.

Reply:
Fixed with [commit](https://github.com/wso2/samples-is/pull/285/commits/92cca01bbfea18fd679f21faa99224aca38f76d1) by continuously writing the user information to CSV file instead of keeping the content in the ArrayNode.